### PR TITLE
add basic typst-docs cmd that spits out json

### PR DESCRIPTION
**Pull request by jcbhmr**
WHY
- generates json which is wayyyy more usable for other tools in other langs
    - ex: using vitepress (js) + dynamic data loading to consume the json output; don't need to manually create a custom provider in rust = contributors don't need to learn rust just to understand how the json gets generated -- just use the builtin default typst-docs cli
    - ex: use github pages builtin jekyll system to use the json output as page content
    - ex: use jq to see what the generated HTML for a page I'm editing (locally!) would look like (the main HTML at least) when published
- lets you do a "cargo check" -like local "does this look good" pass to make sure that all the magic $thing links and image references _actually resolve_ so that it doesn't error in prod when trying to generate the official site
- solves the "waitaminute... how do I actually get some kind of sensible output??" (you don't have to create your own rust crate just to see the rendered docs lol) #2089 comment is a good example of confusion
- basically: you don't have to create a full rust project just to get the json repr of the pages lol
- is VERY TINY so hopefully minimal maintainer burden in addition to regular official closed-source docs

this pr stems from #3425

### CODE REVIEW

**Comment by laurmaedje** [comment]
I don't have a real problem with having this (as you said it's pretty small), but I also don't really understand the motivation. When you are building your own docs renderer, writing the ~80 lines of Rust to interface with typst-docs (which can easily be used as a git dependency) is a minor step and one where you can customize the generation process to your liking with regards to asset placement etc and do extra stuff like search indexing. The Rust interface of `typst-docs` with the `Resolver` already presents a "minimal" public API for retrieving the docs and is less opinionated than a JSON CLI.

Regarding your points:

> lets you do a "cargo check" -like local "does this look good" pass to make sure that all the magic $thing links and image references actually resolve so that it doesn't error in prod when trying to generate the official site

That's actually handled by `cargo test` because there is a `#[test]` in typst-docs. This also happens in typst/typst CI.

> don't need to manually create a custom provider in rust = contributors don't need to learn rust just to understand how the json gets generated

> solves the "waitaminute... how do I actually get some kind of sensible output??"

> you don't have to create a full rust project just to get the json repr of the pages lol

Is it really a common thing to want to look at that JSON though? The JSON doesn't really qualify as particularly sensible output in my book. Pretty much the only two cases I can think of is manually inspecting the HTML (for which it will be a pretty bad experience, I'd rather say to people that it's not yet possible) or building your own docs frontend, where writing your own wrapper will be a tiny portion of the work.

So, in short: I'm not really against it, but am also not sure how useful it really is. If you want to build your own docs frontend that's cool, but having this code in your own repo will actually give you more flexibility for a pretty small amount of code.

**Comment by jcbhmr** [comment]
i think it comes down to whether or not these points are worth the extra code:
1. get a ready-made real-world example of how to use the `Resolver` trait in your own code to actually render the examples and also how to use the other functions
2. dont _have to_ crate a `cargo new my-resolver` project and know how to use `Resolver` in order to get useful assets & json output
3. have a way to language-agnostically output the docs content & can more easily pipe that data to non-Rust tooling

is that worth it? its up to you. i think it is. 🤷‍♂️

my use case is to have multiple submodules of `typst` like `es/` and `zh/` with a `patches/*.patch` so that i have "forks" of typst living in the same repo. then i need to get the json contents of the docs to then pipe that to https://vitepress.dev/ or https://www.getzola.org/. The repeated code across (so far just two) folders was a "hmmm maybe this json generation (yes; it had to be inside the submodule so it would use the forked copy of typst-docs & typst) is generalizable and could be upstreamed so that i dont have to have the same code in two `zh.patch` and `es.patch` files. that's how we ended up here 😆 -- i was basically doing this json generation and then this:

```js
// [path].paths.ts (sibling to '[path].md' template which uses the data)
export default {
  async paths() {
    const rootPagesText = await readFile("pages.json", "utf8")
    const rootPages = JSON.parse(rootPagesText)
    const allPages = rootPages.flatMap(function f(p) {
      return [p, ...p.children.flatMap(f)]
    })
    return allPages.map(page => ({ params: { page } }))
  }
}
```
https://vitepress.dev/guide/routing#dynamic-routes

...and i figured that the json-generation part was generalizable and widely applicable enough to be relevant. it has the added bonus of being the "how to use Resolver in typst-docs" example 🤷‍♂️

File: crates/typst-docs/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    verbose: bool,
+}
+impl<'a> Resolver for MyResolver<'a> {
+    fn commits(&self, from: &str, to: &str) -> Vec<typst_docs::Commit> {
+        if self.verbose {
+            eprintln!("commits({from}, {to})");
+        }
+        vec![]
+    }
+    fn example(
```
Comment (laurmaedje) [inline crates/typst-docs/src/main.rs:22]: Please add newlines between the methods and the struct & impl.

File: crates/typst-docs/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
+use std::{
+    fs::{create_dir_all, write},
+    path::{Path, PathBuf},
+};
```
Comment (laurmaedje) [inline crates/typst-docs/src/main.rs:4]:
```suggestion
use std::fs;
use std::path::{Path, PathBuf};
```
and then `fs::create_dir_all` and `fs::write` (in particular the latter is easier to understand that way in my opinion).

File: crates/typst-docs/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+use clap::Parser;
+use typst::{model::Document, visualize::Color};
```
Comment (laurmaedje) [inline crates/typst-docs/src/main.rs:7]:
```suggestion
use typst::model::Document;
use typst::visualize::Color;
```

File: crates/typst-docs/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+/// JSON to rewrite links and other things to match your site's structure.
+#[derive(Parser, Debug)]
+#[command(version, about, long_about = None)]
+struct Args {
+    /// The generation process can produce additional assets. Namely images. This
+    /// option controls where to spit them out. It assumes that this is the base
+    /// folder of the site ("/" in the URL). Images & example renderings will be
+    /// placed in a folder called "assets/docs" under this directory and expected
+    /// to be made available at "/assets/docs/*".
+    #[arg(short, long, default_value = "_site")]
```
Comment (laurmaedje) [inline crates/typst-docs/src/main.rs:86]: I think `_site` is a misleading name because it just spits out json. Maybe the option can be called `asset_path`?

File: crates/typst-docs/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            eprintln!("link({link})");
+        }
+        None
+    }
+}
+
+/// Generates the JSON representation of the documentation. This can be used
+/// to generate the HTML yourself. You are encouraged to post-process the generated
+/// JSON to rewrite links and other things to match your site's structure.
+#[derive(Parser, Debug)]
```
Comment (laurmaedje) [inline crates/typst-docs/src/main.rs:78]: Please add a note here than the JSON structure may change at any time.

**Review by laurmaedje — COMMENTED** [review]
Okay, I'm fine with having this, just a few comments.

File: crates/typst-docs/src/main.rs (hunk 4 above)
Comment (jcbhmr) [reply crates/typst-docs/src/main.rs]: i changed it from `--out-dir` and `${out_dir}/assets/docs/*` and default of `_site` to `--assets-dir` and `${assets_dir}/*` and default to `assets`.

> The same sort of goes for the root path of the docs themselves, but it's harder to make the /docs prefix optional as it's hardcoded in more places.

I've found that this works as a hack for the multi-lang docs hosting thing:

```js
// es/[page].paths.ts
let text = await readFile(
  new URL(import.meta.resolve("./typst-docs.json")),
  "utf8"
);
text = text.replaceAll(/(?<=[^\w\-])\/docs\//g, "/es/");
text = text.replaceAll(/(?<=[^\w\-])\/assets\/docs\//g, "/es/assets/docs/");
const rootPages = JSON.parse(text) as PageModel[];
```

so I could imagine doing the reverse:

```rs
let json = ...;
// yeah, yeah.. regex find/replace on raw json is brittle but it works lol
let json = Regex::new("(?<=[^\w\-])/docs/").replace_all(json, "/");
```

**Comment by jcbhmr** [comment]
there WERE matches in the naive "/docs/" => "/" replacement
image

there are NO matches in the regex  `([^\w\-])/docs/`
image

**Comment by laurmaedje** [comment]
If we want to support movable root dir, we should at least look into doing it directly in the generation process (as pointed out in the fixme).

**Comment by jcbhmr** [comment]
would you like me to:
1. revert the `([^\w\-])/docs/` hack and not do anything else
2. revert the `([^\w\-])/docs/` hack and try to find/code a way to allow a user to change the `/docs/` base url in this pr
3. keep the `([^\w\-])/docs/` hack and try to pursue a way to allow a user to change the `/docs/` base url in another pr
4. or something else?

**Comment by jcbhmr** [comment]
i tracked down all matches for "/docs/" using vscode search and got a few options working:

1. use a BASE_URL fallback to BASE_PATH fallback to "/docs/" global lazy static var
2. add a `base(&self) -> Option<&str>` `Resolver` method getter and fall back to `/docs/` called MULTIPLE TIMES wherever needed
3. add a `base(&self) -> Option<&str>` `Resolver` method getter and fall back to `/docs/` called ONCE and passed via params to all that need it

other options that are viable but I didn't explicitly try out yet:
1. add a second `provide(resolver, base)` parameter and pass it via param drilling to all that need it
2. use a mutable global BASE var that can be mutated from outside the crate

**Comment by laurmaedje** [comment]
I like 2 best. I would adjust it to make it `base(&self) -> &str` and make it required.

**Comment by jcbhmr** [comment]
> try to find/code a way to allow a user to change the /docs/ base url in this pr if it's reasonable simple (I'm not entirely sure) and if you come to the conclusion that it's not easy to do, we can keep the hack.

i came to a compromise of splitting it into two prs. why?
1. so that "how should the user specify a base URL? env var? extra param? param drilling? call `base()` once? multiple times?" can all happen in https://github.com/typst/typst/pull/3484
2. so that this or that can be merged faster since its easier to ✅ approve (hopefully) 😊

also just so that I can mentally separate these two things; even though they are closely related they are separate changes in my brain 🧠 🤷‍♀️

this pr was force-pushed back to WITH THE HACK.

File: crates/typst-docs/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+use clap::Parser;
+use regex::Regex;
+use typst::model::Document;
+use typst::visualize::Color;
+use typst_docs::{provide, Html, Resolver};
+use typst_render::render;
+
+#[derive(Debug)]
+struct MyResolver<'a> {
```
Comment (laurmaedje) [inline crates/typst-docs/src/main.rs:12]:
```suggestion
struct CliResolver<'a> {
```

File: crates/typst-docs/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+        if self.verbose {
+            eprintln!("link({link})");
+        }
+        None
+    }
+}
+
+/// Generates the JSON representation of the documentation. This can be used to
+/// generate the HTML yourself. You are encouraged to post-process the generated
+/// JSON to rewrite links and other things to match your site's structure. Be
```
Comment (laurmaedje) [inline crates/typst-docs/src/main.rs:83]: I think we can remove this encouragement for post-processing now with the new `base` config.

**Review by laurmaedje — COMMENTED** [review]
The other PR is merged, so this can be rebased/main-merged-into-it. I've left two other minor comments and then this should be ready for merging.

**Comment by laurmaedje** [comment]
Thanks!
